File: src/a3_mathias/Alignment.cpp
#include "Alignment.hpp"
#include <utility>
#include <stdexcept>
#include <algorithm>
#include <iostream> // Debug
#include <iomanip> // Debug


Alignment::Alignment(const std::string& seq_v, const std::string& seq_h) {
    this->seqv = seq_v;
    this->seqh = seq_h;
}

void Alignment::compute(const int match, const int mismatch, const int gap, const bool local_align) {
    computeCalled = true;
    f.clear();
    t.clear();

    uint32_t width = seqh.size()+1;
    uint32_t height = seqv.size()+1;
    long maxScore = 0;
    int matchScore = 0;


    f.resize(width, std::vector<long>(height));
    t.resize(width, std::vector<Traceback>(height));
    //std::cout << "Matrix Borders:\n " << seqh << "\n"; // DEBUG
    //for (uint32_t i = 0; i < seqv.size(); i++) std::cout << seqv[i] << "\n"; // DEBUG

    // Initialization
    f[0][0] = 0;
    t[0][0] = Traceback::NONE;

    for (uint32_t i = 0; i < width; i++) {
        f[i][0] = static_cast<long>(i) * gap;
        t[i][0] = Traceback::HORIZONTAL;
    }
    for (uint32_t j = 0; j < height; j++) {
        f[0][j] = static_cast<long>(j) * gap;
        t[0][j] = Traceback::VERTICAL;
    }

    // Recurrence

    for (uint32_t i = 1; i < width; i++) {
        for (uint32_t j = 1; j < height; j++) {
            matchScore = (seqh[i-1] == seqv[j-1]) ? match : mismatch;

            maxScore = f[i][j-1] + gap;
            t[i][j] = Traceback::VERTICAL;


            if (f[i-1][j] + gap > maxScore) {
                maxScore = f[i-1][j] + gap;
                t[i][j] = Traceback::HORIZONTAL;
            }
            if (f[i-1][j-1] + matchScore >= maxScore) {
                maxScore = f[i-1][j-1] + matchScore;
                t[i][j] = Traceback::DIAGONAL;
            }
            

            f[i][j] = maxScore;

        }
    }

    score = f[width-1][height-1];



    // Debug: Print matrix
    // for (uint32_t j = 0; j < height; j++) {
    //     for (uint32_t i = 0; i < width; i++) {
    //         std::cout << "[";
    //         if (t[i][j] == Traceback::DIAGONAL) std::cout << "\\";
    //         else if (t[i][j] == Traceback::HORIZONTAL) std::cout << "<";
    //         else if (t[i][j] == Traceback::VERTICAL) std::cout << "|";
    //         else if (t[i][j] == Traceback::NONE) std::cout << "X";
    //         std::cout << "" << std::setw(4) << f[i][j] << "] ";
    //     }
    //     std::cout << "\n";
    // }
}

int Alignment::getScore() const {
    if (!computeCalled) throw(std::runtime_error("Compute hasn't been called!"));
    return score;
}


void Alignment::getAlignment(std::string& a1, std::string& gaps, std::string& a2) const {
    if (!computeCalled) throw(std::runtime_error("Compute hasn't been called!"));

    a1 = "";
    a2 = "";
    gaps = "";
    uint32_t i = seqh.size();
    uint32_t j = seqv.size();

    while (!(i == 0 && j == 0)) {
        if (t[i][j] == Traceback::DIAGONAL) {
            a1 += seqv[j-1];
            a2 += seqh[i-1];
            gaps += (seqv[j-1] == seqh[i-1]) ? "|" : " ";
            i--; j--;
        }
        else if (t[i][j] == Traceback::HORIZONTAL) {
            a1 += "-";
            a2 += seqh[i-1];
            gaps += " ";
            i--;
        }
        else {
            a1 += seqv[j-1];
            a2 += "-";
            gaps += " ";
            j--;
        }
    }

    std::reverse(a1.begin(), a1.end());
    std::reverse(gaps.begin(), gaps.end());
    std::reverse(a2.begin(), a2.end());

    //std::cout << a1 << "\n" << gaps << "\n" << a2 << "\n";
}

File: src/a3_mathias_Rust/src/alignment/alignment.rs
/*
This is an implementation of the Needleman-Wunsch algorithm (aufgabe3) in Rust.
I did this mainly for the practice, and also to see the differences in performance
between a near identical implementation in C++ and Rust.
*/

/*
Tracebacks:
'0': None
'd': Diagonal
'v': Vertical
'h': Horizontal
*/

pub struct Alignment {
    seqv: String,
    seqh: String,
    f: Vec<Vec<i32>>,
    t: Vec<Vec<char>>,
    score: i32
}

impl Alignment {
    pub fn new(horizontal_sequence: &str, vertical_sequence: &str) -> Alignment {
        Alignment {
            seqh: String::from(horizontal_sequence),
            seqv: String::from(vertical_sequence),
            f: Vec::new(),
            t: Vec::new(),
            score: 0
        }
    }

    // Compute the alignment
    pub fn compute(&mut self, match_score: i32, mismatch_score: i32, gap: i32) {
        // Todo: Compute called?
        self.f.clear();
        self.t.clear();

        let width = self.seqh.len()+1;
        let height = self.seqv.len()+1;

        let mut current_score: i32;
        let mut max_score : i32;

        self.f.resize(width, vec![0; height as usize]);


        self.t.resize(width, vec!['0'; height as usize]);

        // Correct the first horizontal row to point towards (0,0)
        for i in 0..width {
            self.t[i][0] = 'h';
            self.f[i][0] = (i as i32)*gap;
        }
        for i in 0..height {
            self.t[0][i] = 'v';
            self.f[0][i] = (i as i32)*gap;
        }

        for row in 1..width {
            for line in 1..height {
                if self.seqh.chars().nth(row-1) == self.seqv.chars().nth(line-1) {
                    current_score = match_score;
                }
                else {
                    current_score = mismatch_score;
                }

                // Start looking for max

                max_score = self.f[row][line-1] + gap;
                self.t[row][line] = 'v';

                if self.f[row-1][line] + gap > max_score {
                    max_score = self.f[row-1][line] + gap;
                    self.t[row][line] = 'h';
                }
                if self.f[row-1][line-1] + current_score >= max_score {
                    max_score = self.f[row-1][line-1] + current_score;
                    self.t[row][line] = 'd';
                }

                self.f[row][line] = max_score;
            }
        }
        self.score = self.f[width-1][height-1];

    }

    // Just a simple function to return the score of the alignment made.
    pub fn get_score(&self) -> i32 {
        self.score
    }

    // Helper function in case we need to view the matrix
    pub fn print_matrix(&self) {
        for row in &self.f {
            for &element in row {
                print!("{} ",element);
            }
            print!("\n");
        }
    }

    // This function will put our results into 3 strings
    pub fn get_alignment(&self, a1: &mut String, gaps: &mut String, a2: &mut String) {
        a1.clear();
        gaps.clear();
        a2.clear();

        let mut i = self.seqh.len();
        let mut j = self.seqv.len();

        while !(i == 0 && j == 0) && (self.t[i][j] != '0') {
            if self.t[i][j] == 'd' {
                a1.push(self.seqv.chars().nth(j-1).unwrap());
                a2.push(self.seqh.chars().nth(i-1).unwrap());
                if self.seqv.chars().nth(j-1).unwrap() == self.seqh.chars().nth(i-1).unwrap() {
                    gaps.push('|');
                } else {
                    gaps.push(' ');
                }
                i -= 1;
                j -= 1;
            }
            else if self.t[i][j] == 'v' {
                a1.push(self.seqv.chars().nth(j-1).unwrap());
                a2.push('-');
                gaps.push(' ');
                j -= 1;
            }
            else {
                a1.push('-');
                a2.push(self.seqh.chars().nth(i-1).unwrap());
                gaps.push(' ');
                i -= 1;
            }
        }

        let mut reverser: Vec<char>;
        reverser = a1.chars().rev().collect();
        a1.clear();
        a1.extend(reverser);

        reverser = a2.chars().rev().collect();
        a2.clear();
        a2.extend(reverser);
        reverser = gaps.chars().rev().collect();
        gaps.clear();
        gaps.extend(reverser);

        println!("{}\n{}\n{}\n", a1, gaps, a2);


    }
}